tests/DocBlock/TypeExpressionTest.php

```diff
@@ -423,7 +436,7 @@ public static function provideSortTypesCases(): iterable
 
         yield 'nullable callable' => [
             '?callable(Foo|Bar): Foo|Bar',
-            '?callable(Bar|Foo): Bar|Foo',
+            '?callable(Bar|Foo): Foo|Bar',
```

Comment:
I'm not sure if we should change expected types' list (which means changing input from one type to union type) or rather add parentheses to input's return type to keep single type 🤔. What is certain, we need to have test cases for both scenarios, and mix of them like:

```php
yield ['\\Closure(float|int): (bool|int)|int', ['\\Closure(float|int): (bool|int)', 'int']];
```

@kubawerlos what do you think?

Reply:
Thanks to the 1st commit union from both sides is tested for all `provideGetTypesCases` cases (so `\\Closure(float|int): (bool|int)` case will be tested with union)

Reply:
Ah, OK 🙂. So there is only one thing to decide: which way we want to fix current cases that started to behave differently after fix:

- add parentheses to them, too keep them as single type
- treat them as buggy cases and change expected output, so now they would ensure correct (but different) behavior

I would go with the former, to keep the original intention, which clearly was sorting union type in the return type. Then, I would add another test cases for sorting without parentheses, which should result in changed order of the types, like in this commented case.

Reply:
There is such case already: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/e4d89945e234226121ded3d9da543a1023b16f43/tests/DocBlock/TypeExpressionTest.php#L468

As long as the sorting is done on AST, and it is, with the parsing tested above, it is all tested.

But feel free to write here what testcases you miss and I can add them, it is the easiest task for me :)

Reply:
In short: I would add parentheses to all modified cases to keep their original intention + copy those cases but without parentheses, and of course with fixed expected output. This way we would have whole current test suite "untouched" (from intention point of view), and extended with new cases that would cover change introduced in this PR. Does it sound good to you?

Reply:
I wanted rather something like this:

```diff
Index: tests/DocBlock/TypeExpressionTest.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/DocBlock/TypeExpressionTest.php b/tests/DocBlock/TypeExpressionTest.php
--- a/tests/DocBlock/TypeExpressionTest.php	(revision 0aa07209dbcb59c751d4619eb5ce559f6a53b34d)
+++ b/tests/DocBlock/TypeExpressionTest.php	(date 1684422571919)
@@ -405,8 +405,13 @@
         ];
 
         yield 'simple in callable return type' => [
-            'callable(): string|float',
-            'callable(): string|float',
+            'callable(): (string|float)',
+            'callable(): (float|string)',
+        ];
+
+        yield 'callable with union return type and within union itself' => [
+            'callable(): (string|float)|bool',
+            'bool|callable(): (float|string)',
         ];
 
         yield 'simple in closure argument' => [
@@ -420,13 +425,23 @@
         ];
 
         yield 'simple in closure return type' => [
-            'Closure(): string|float',
-            'Closure(): string|float',
+            'Closure(): (string|float)',
+            'Closure(): (float|string)',
+        ];
+
+        yield 'closure with union return type and within union itself' => [
+            'Closure(): (string|float)|bool',
+            'bool|Closure(): (float|string)',
         ];
 
         yield 'with multiple nesting levels' => [
-            'array{0: Foo<int|bool>|Bar<callable(string|float|array<int|bool>): Foo|Bar>}',
-            'array{0: Bar<Bar|callable(array<bool|int>|float|string): Foo>|Foo<bool|int>}',
+            'array{0: Foo<int|bool>|Bar<callable(string|float|array<int|bool>): (Foo|Bar)>}',
+            'array{0: Bar<callable(array<bool|int>|float|string): (Bar|Foo)>|Foo<bool|int>}',
+        ];
+
+        yield 'with multiple nesting levels and callable within union' => [
+            'array{0: Foo<int|bool>|Bar<callable(string|float|array<int|bool>): (Foo|Bar)|Baz>}',
+            'array{0: Bar<Baz|callable(array<bool|int>|float|string): (Bar|Foo)>|Foo<bool|int>}',
         ];
 
         yield 'nullable generic' => [
@@ -435,13 +450,15 @@
         ];
 
         yield 'nullable callable' => [
-            '?callable(Foo|Bar): Foo|Bar',
-            '?callable(Bar|Foo): Foo|Bar',
+            '?callable(Foo|Bar): (Foo|Bar)',
+            '?callable(Bar|Foo): (Bar|Foo)',
         ];
 
-        yield 'nullable callable and union' => [
-            '?callable(Foo|Bar): (Foo|Bar)|Bar',
-            '?callable(Bar|Foo): (Bar|Foo)|Bar',
+        // This union type makes no sense in general (it should be `Bar|callable|null`)
+        // but let's ensure nullable types are also sorted.
+        yield 'nullable callable with union return type and within union itself' => [
+            '?callable(Foo|Bar): (Foo|Bar)|?Bar',
+            '?Bar|?callable(Bar|Foo): (Bar|Foo)',
         ];
 
         yield 'nullable array shape' => [
```

You can apply this patch or I can commit and push it to this PR.